Fixes #30473 - UI for Global Registration

stejskalleos opened the pull request:
User interface for Global Registration feature, allowing users to quickly generate command via simple form.

*Features*
* New page _Register Host_ (Go to _Hosts > All hosts_ and click on **Register Host** button)
* Form extendable from plugins by `extend_allowed_registration_vars :your_param`

![registration_UI_new](https://user-images.githubusercontent.com/59385976/95834910-ddc87480-0d3d-11eb-8a0e-85e0ac3f3031.png)

---

PR is part of **Simple & automatic host registration WF**, see:
* [RFC](https://community.theforeman.org/t/rfc-simple-automatic-host-registration-wf/19588)
* [Redmine (this PR)](https://projects.theforeman.org/issues/30473)
* [Redmine (Parent task)](https://projects.theforeman.org/issues/30440)


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

stejskalleos committed:
Fixes #30473 - UI for Global Registration

New page 'All Hosts > Register Host' with form for
easy generation of command for Global Registration endpoint.

Form allows user to define parameters required for the
Global Registration endpoint, plus extended parameters
from plugins, for example ':activation_key' from Katello plugin.

ekohl commented:
First impression from the screenshot:
* Proxy is an ambiguous term. I think this should be Smart Proxy.
* JWT expiration needs a unit. 4 doesn't tell me. From the code I think I remember it's hours but we can't expect users to know. Perhaps also a number field so the browser accepts only integers?
* I don't understand what `Param From Rex` would do.
* Some help text describing the flow can be useful. If I press submit, I don't know what would happen.

ekohl commented on app/controllers/registration_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    if params['smart_proxy'].empty?
+      global_registration_url
+    else
+      "#{SmartProxy.authorized(:view_smart_proxies).find(params[:smart_proxy]).url}/register"
```
I'm always a fan of defensive programming. Don't assume `SmartProxy.authorized(:view_smart_proxies).find(params[:smart_proxy])` actually returns an object. The Smart Proxy may have been deleted since the form was rendered and it'll return nil, causing an internal server error.

I also don't like hardcoding URLs. Perhaps there should be a helper for it? The helper can return a valid SmartProxy object (with the register feature) or raise a proper exception

ekohl commented on app/controllers/registration_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    @locations = Location.authorized(:view_locations).select(:id, :name)
+    @host_groups = Hostgroup.authorized(:view_hostgroups).select(:id, :name)
+    @operating_systems = Operatingsystem.authorized(:view_operatingsystems).select(:id, :title)
+    @smart_proxies = SmartProxy.authorized(:view_smart_proxies).select(:id, :name)
```
This should be only smart proxies with the Register feature (depends on https://github.com/theforeman/foreman/pull/8024). IMHO you shouldn't add Smart Proxy support in this page without that. My suggestion would be to leave out the Smart Proxy altogether until the Register module support is actually there.

ekohl commented on app/controllers/registration_controller.rb:
```diff
@@ -0,0 +1,53 @@
+                         hostgroup_id: params[:host_group],
+                         operatingsystem_id: params[:operating_system],
+                         url: url,
+                       }).delete_if { |_, v| v.empty? }
```
Perhaps it should use `blank?` so you can also deal with `nil`?
```suggestion
                       }).delete_if { |_, v| v.blank? }
```

ekohl commented on app/views/registration/_form.erb:
```diff
@@ -0,0 +1,74 @@
+      $(aux).html(text);
+      document.body.appendChild(aux);
+      aux.select();
+      document.execCommand("copy");    <%= submit_tag _("Submit"), :class => "btn btn-primary" %>
+    <%= link_to _("Cancel"), hosts_path, :class => "btn btn-default" %>
```
These tags look odd within javascript Copy/paste error?

ekohl commented on app/views/registration/_form.erb:
```diff
@@ -0,0 +1,74 @@
+    <div class="form-group ">
+      <label class="col-md-2 control-label"><%= _('Command') %></label>
+      <div class="col-md-8">
+        <pre class="ellipsis white-space-normal" id="registration_command">curl -X GET "<%= @endpoint %>?<%= @registration_args.to_query %>" <%= @headers %> | bash</pre>
```
Nit: should the controller provide a complete URL? IMHO it's not something you'd do in a view

ekohl commented on app/views/registration/_form.erb:
```diff
@@ -0,0 +1,74 @@
+    </div>
+  <% end %>
+<% end %>
+
+
```
I'm allergic to empty lines at the end

ekohl commented on app/controllers/registration_controller.rb:
```diff
@@ -0,0 +1,53 @@
+
+  def registration_args
+    from_plugins = Hash[plugin_params.map { |key| [key, params[key]] }]
+    url = params['smart_proxy'].empty? ? '' : SmartProxy.authorized(:view_smart_proxies).find(params[:smart_proxy]).url
```
Why do you need to pass the URL as a registration arg? I thought the `/register` on the Smart Proxy would ensure the correct template is provided and there is no need to do this on the Foreman side.

stejskalleos replied:
See: https://github.com/theforeman/foreman/pull/8054

ekohl commented on app/views/registration/_form.erb:
```diff
@@ -0,0 +1,71 @@
+    </div>
+  </div>
+  <div class="form-group ">
+    <label class="col-md-2 control-label"><%= _('JWT Expiration') %>&nbsp;<%= popover('', _('Expiration of JWT authorization token in hours'), options = {}) %></label>
+    <div class="col-md-4">
+      <%= number_field_tag 'jwt_expiration', 4, class: 'form-control', min: 1 %>
+    </div>
```
I do believe bootstrap can have a suffix field where you can place the `hours` bit, but it's been a while since I did any UI work.

stejskalleos replied:
I didn't find anything, can you give me a link?

timogoebel commented on app/views/registration/_form.erb:
```diff
@@ -0,0 +1,71 @@
+    </div>
+  </div>
+  <div class="form-group ">
+    <label class="col-md-2 control-label"><%= _('JWT Expiration') %>&nbsp;<%= popover('', _('Expiration of JWT authorization token in hours'), options = {}) %></label>
```
Let's label this more user friendly: "Link expiration" or something. The JWT is an implementation detail in my opinion and a user should not care about this.

ekohl replied:
Good suggestion, but I'd also consider `token expiration`.

ares replied:
Even more precise and descriptive would be `Token lifetime` or `Token validity`, given it's relative value in hours. Or `Token expires in`.

ares commented on app/views/registration/_form.erb:
```diff
@@ -0,0 +1,71 @@
+<% title _("Register Host") %>
+<%= form_tag({controller: 'registration', action: 'create' }, method: :post, class: 'form-horizontal well') do %>
```
I guess we can't easily use traditional form helpers, since we don't have model here.

ares commented on app/views/registration/_form.erb:
```diff
@@ -0,0 +1,71 @@
+    </div>
+  <% end%>
+  <div class="form-group ">
+    <%= submit_tag _("Submit"), :class => "btn btn-primary" %>
```
Could we label the submit button with "Generate the registration command"? If it's too long, then "Generate command", you'll see how it looks and what's better.

ares commented on app/views/registration/create.html.erb:
```diff
@@ -0,0 +1 @@
+<%= render partial: 'form' %>
```
why not `<%= render 'form' %>` like you do in new.html.erb?

ares reviewed (commented):
few additional comments, one of the PR is in, this now waits on https://github.com/theforeman/smart-proxy/pull/768 only (and many suggestions here)

stejskalleos commented:
> JWT expiration needs a unit. 4 doesn't tell me.

It was described in help, moved to label.


> I don't understand what Param From Rex would do.

It's extension from plugin. For katello there will be `activation_key` for example.

> Some help text describing the flow can be useful. If I press submit, I don't know what would happen.

Renamed to `Generate command`

ares commented on app/controllers/registration_controller.rb:
```diff
@@ -0,0 +1,52 @@
+  end
+
+  def endpoint
+    return global_registration_url if params['smart_proxy'].empty?
```
just to be a bit more defensive, `blank?` would make more sense, e.g. if in future the smart proxy selector could be disabled

ares commented on app/controllers/registration_controller.rb:
```diff
@@ -0,0 +1,52 @@
+    if proxy
+      "#{proxy.url}/register"
+    else
+      global_registration_url
```
Should this instead raise an error? If we got here, user passed a proxy id for which he or she does not have view persmissions on. I think it should cause some exception.

ares commented on app/views/registration/_form.erb:
```diff
@@ -0,0 +1,71 @@
+    <div class='form-group '>
+      <label class='col-md-2 control-label'><%= pr.to_s.titleize %></label>
+      <div class='col-md-4'>
+        <%= text_field_tag pr, params[pr], class: 'form-control' %>
```
This does not allow any advanced customization of the form. E.g. Katello certainly will want to add activation key selector, text field is not the best user experience. I'm fine with this for an initial version, but I think it would be good to add a pagelet instead and let plugins define whatever form field they find suitable.

ares commented on app/views/registration/_form.erb:
```diff
@@ -0,0 +1,71 @@
+
+  <% if action_name == 'create' %>
+  <script type='text/javascript'>
+    function register_copy(text) {
```
I'm almost certain this is not how JS should be used these days :-) I see we have it in registration tab already, however I think it is worth extracting this to our tfm node module and pass just the text there. The function is 1:1 to https://github.com/theforeman/foreman/blob/develop/app/views/hosts/_registration_tab.html.erb#L1-L10

ares reviewed (changes requested):
I haven't tested yet, but leaving few code related comments. I don't insist on pagelets in this PR but all the rest seems important to change here. Also one question - how much work it would be to do the same in react? Why you decided to use ERB here? Do we have all the helpers necessary for this view in react already? I don't insist, however I know we have slotnfill extension mechanism there and I was under the impression, we already have form builders available.

ares reviewed (changes requested):
Thanks for great functionality. During testing I found these things that I'd prefer to be improved:

1) when I enter the registration page, I'm in org 2 but the org selector defaults to org 1, could we set the default to `Organization.current`? The same for location please.

2) the hostgroup (and other resources) selectors respect the current context (Organization.current and Location.current). It does not make sense to display the org and loc selectors in that case. If I'm in org 2 but chose org 1, I can still select host group from org 2 only, causing a resource mismatch and failing registration.

I suggest if user is in a specific org, don't render the select box but hidden instead with `Organization.current.id` as a value. If Organization.current is nil, it means use is in any context and then we should show the org selector. The same for locations.

3) If the url contains `&`, shell has problems with it. We either need to escape the URL or (I'd prefer this) wrap the url in single quotes `'`.

stejskalleos commented:
PR updated
* Removed selects for Organizations and Locations (fixed taxonomy mismatch issue)
* Added support for pagelets, now easily extendable from plugins:
```
        extend_page "registration/_form" do |cx|
          cx.add_pagelet :global_registration, :name => N_("My name"), :partial => "my_view", :priority => 100, :id => 'registration-rex'
        end
```
* Fixed minor issues

ares reviewed (approved):
Tested, all works as expected, ready for merge pending tests turning green.

ares commented:
Thanks @stejskalleos, merging!
